packages/beacon-node/src/chain/seenCache/seenGossipBlockInput.ts

```diff
@@ -350,6 +361,100 @@ export class SeenGossipBlockInput {
+    const blockCache = this.blockInputCache.get(blockRoot);
+    if (blockCache === undefined || blockCache.fork !== ForkName.fulu) {
+      return;
+    }
+
+    // If already have all columns, exit
+    if (
+      blockCache.cachedData &&
+      (blockCache.cachedData.fork !== ForkName.fulu ||
+        this.hasSampledDataColumns(blockCache.cachedData.dataColumnsCache))
+    ) {
+      return;
+    }
+
+    // Process KZG commitments into versioned hashes
+    let versionedHashes: Uint8Array[];
+
+    if (blockCache.block) {
+      const block = blockCache.block as fulu.SignedBeaconBlock;
+      versionedHashes = block.message.body.blobKzgCommitments.map(kzgCommitmentToVersionedHash);
+    } else if (blockCache.cachedData) {
+      const firstSidecar = blockCache.cachedData.dataColumnsCache.values().next().value;
+      if (!firstSidecar) {
+        return;
+      }
+
+      versionedHashes = firstSidecar.dataColumn.kzgCommitments.map(kzgCommitmentToVersionedHash);
+    } else {
+      throw new Error("blockInputCache missing both block and cachedData");
+    }
+
+    // Return if block has no blobs
+    if (versionedHashes.length === 0) {
+      return;
+    }
+
+    // Get blobs from execution engine
+    const blobs = await this.executionEngine.getBlobs(blockCache.fork, versionedHashes);
+
+    // Execution engine was unable to find one or more blobs
+    if (blobs === null) {
+      return;
+    }
+
+    // Return if we received all data columns while waiting for getBlobs
+    if (blockCache.cachedData && this.hasSampledDataColumns(blockCache.cachedData.dataColumnsCache)) {
+      return;
+    }
+
+    let dataColumnSidecars: fulu.DataColumnSidecars;
+    const cellsAndProofs = getCellsAndProofs(blobs);
+    if (blockCache.block) {
+      dataColumnSidecars = getDataColumnSidecarsFromBlock(
+        config,
+        blockCache.block as fulu.SignedBeaconBlock,
+        cellsAndProofs
+      );
+    } else if (blockCache.cachedData) {
+      const firstSidecar = blockCache.cachedData.dataColumnsCache.values().next().value;
+      if (!firstSidecar) {
+        return;
+      }
+      dataColumnSidecars = getDataColumnSidecarsFromColumnSidecar(firstSidecar.dataColumn, cellsAndProofs);
+    } else {
+      throw new Error("blockInputCache missing both block and cachedData");
+    }
+
+    // Publish columns if and only if subscribed to them
+    const sampledColumns = this.custodyConfig.sampledColumns.map((columnIndex) => dataColumnSidecars[columnIndex]);
+
+    this.emitter.emit(ChainEvent.publishDataColumns, sampledColumns);
+
+    for (const column of sampledColumns) {
+      this.getGossipBlockInput(
+        config,
+        {
+          type: GossipedInputType.dataColumn,
+          dataColumnSidecar: column,
+          dataColumnBytes: null,
+        },
+        // TODO: Pass in metrics. Should this use a different availability source?
+        null
+      );
+    }
```

Comment:
I do not think you should call `getGossipBlockInput`.  I think you should just add the columns to the blockInput and then resolve the dataAvailability promise to unblock all of the consumers.